zerver/lib/markdown/__init__.py

```diff
@@ -2083,15 +2083,19 @@ def make_md_engine(realm_filters_key: int, email_gateway: bool) -> None:
 # function on the URLs; they are expected to be HTML-escaped when
 # rendered by clients (just as links rendered into message bodies
 # are validated and escaped inside `url_to_a`).
-def topic_links(realm_filters_key: int, topic_name: str) -> List[str]:
-    matches: List[str] = []
+def topic_links(realm_filters_key: int, topic_name: str) -> List[Dict[str, Union[str, int]]]:
+    matches: List[Dict[str, Union[str, int]]] = []
 
     realm_filters = realm_filters_for_realm(realm_filters_key)
 
     for realm_filter in realm_filters:
         pattern = prepare_realm_pattern(realm_filter[0])
         for m in re.finditer(pattern, topic_name):
-            matches += [realm_filter[1] % m.groupdict()]
+            matched_details = m.groupdict()
+            matched_text = matched_details["linkifier_actual_match"]
+            matches += [dict(url=realm_filter[1] % matched_details,
```

Comment:
I've previously thought this construction isn't very easy to read. In fact, even when reading this code yesterday, I glanced over it as code that somehow works but I'm not interested in looking up the specifics of how the regex match API works, while _all it is is a format string_.

I'm not sure how others feel about it but I feel we should have a follow-up commit to explicitly say what we're doing, like:

```python
# format the realm_filter's url string using the matched text.
url = realm_filter[1] % match_details
```

Reply:
I was also confused as to what was going on here the first time I read it. I added a similar comment and one below as to why we are including the match_text. Thanks for pointing it out :)